docs/contributing/Feature-flags.md

```diff
@@ -0,0 +1,328 @@
+     @override
+     bool get isUnicornEmojisEnabled => isEnabled(unicornEmojis);
+   }
+   ```
+
+   [`FlutterFeatureFlagsIsEnabled`]: ../../packages/flutter_tools/lib/src/flutter_features.dart
+
+### Allowing flags to be enabled
+
+By default, after [adding a flag](#adding-a-flag), the flag is considered
+_disabled_, and _cannot_ be enabled outside of our own [unit tests](#tests).
+This allows iterating locally with the code without having to support users or
+field issues related to the flag.
+
+After some time, you may want to allow the flag to be enabled.
+
+Using the options `master`, `beta` or `stable`, you can make the flag
+configurable in those channels. For example, to make the flag available to be
+enabled (but still off by default) on the `master` channel:
+
+```dart
+const Feature unicornEmojis = Feature(
+  name: 'add unicorn emojis in lots of fun places',
+  configSetting: 'enable-unicorn-emojis',
+  master: FeatureChannelSetting(available: true),
+);
+```
+
+Or to make it available on all channels:
+
+```dart
+const Feature unicornEmojis = Feature(
+  name: 'add unicorn emojis in lots of fun places',
+  configSetting: 'enable-unicorn-emojis',
+  master: FeatureChannelSetting(available: true),
+  beta: FeatureChannelSetting(available: true),
+  stable: FeatureChannelSetting(available: true),
+);
+```
+
+### Enabling a flag by default
+
+Once a flag is ready to be enabled by default, once again it can be configured
+on a per-channel basis.
+
+For example, enabled on `master` by default, but disabled elsewhere:
+
+```dart
+const Feature unicornEmojis = Feature(
+  name: 'add unicorn emojis in lots of fun places',
+  configSetting: 'enable-unicorn-emojis',
+  master: FeatureChannelSetting(available: true, enabledByDefault: true),
+  beta: FeatureChannelSetting(available: true),
+  stable: FeatureChannelSetting(available: true),
+);
+```
+
+Once the flag is ready to be enabled in every environment:
+
+```dart
+const Feature unicornEmojis = Feature.fullyEnabled(
+  name: 'add unicorn emojis in lots of fun places',
+  configSetting: 'enable-unicorn-emojis',
+);
+```
+
+### Removing a flag
+
+After a flag is no longer useful (perhaps the experiment has concluded, the
+flag has been enabled successfully for 1+ stable releases), _most_[^1] flags
+should be removed so that the older behavior (or lack of a feature) can be
+refactored and removed from the codebase, and there is less of a possibility of
+conflicting flags.
+
+To a remove a flag, follow the opposite steps of
+[adding a flag](#adding-a-flag).
+
+You may need to remove references to the (disabled) flag from unit or
+integration tests as well.
+
+[^1]: Some flags might have a longer or indefinite lifespan, but this is rare.
+
+## Using a flag to drive behavior
+
+Once you have a flag, you can use it to conditionally enable something or
+provide a different execution branch.
+
+### Tool
+
+Feature flags can be accessed either by adding (and providing) an explicit
```

Comment:
Nit: Maybe be redundant here and say "In the Flutter tool, feature flags...". I'm thinking to avoid people skimming this and copying this code to the wrong place.

Same for below in the framework section.